android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -165,13 +166,18 @@ class GeckoEngineSession(
         parent: EngineSession?,
         flags: LoadUrlFlags,
         additionalHeaders: Map<String, String>?,
+        shouldKeepSearchTerm: Boolean,
     ) {
         val scheme = Uri.parse(url).normalizeScheme().scheme
         if (BLOCKED_SCHEMES.contains(scheme) && !shouldLoadJSSchemes(scheme, flags)) {
             logger.error("URL scheme not allowed. Aborting load.")
             return
         }
 
+        if (!shouldKeepSearchTerm) {
+            notifyObservers { onSearchTermDismiss() }
```

Comment:
I don't think it's right for the browser engine implementation to control dismissing search terms. It has otherwise no notion of search terms so this isn't quite the right abstraction. 

I think this should be handled outside the engine by the caller and/or the observers of these events. I would propose adding additional sate to the load request (in the browser store) for this. The state you added here is whether or not to keep the search terms, which can properly be reduced further to whether or not the current/last load is a search. Once in the store you can access it everywhere and dismiss as needed.

Reply:
In other words, I think we can introduce more generic events from the engine if needed, add some state to the store and handle it in the caller / observers. Happy to talk through options! Sorry that this is a bit of a painful patch to go through.

Reply:
A perhaps even simpler solution is to dispatch an action to clear the search terms  (`UpdateSearchTermsAction`) in the use cases directly. Then you also don't need to loop through this additional parameter. In the use cases, it's implicitly clear if the request is a search or not so it's easy to decide there and they already dispatch actions to the store.

We can then keep `onNavigateBack` and a add a new `onNavigateForward` and dispatch the action as well in `EngineObserver`. This makes the diff much smaller.

Then the only case missing is line 219 below where you dispatch as part of loadData, but we can introduce a new event for that so it's not tied to search terms.